kie-server-parent/kie-server-api/src/main/java/org/kie/server/api/KieServerEnvironment.java

```diff
@@ -27,10 +27,20 @@
     private static final Pattern VERSION_PAT = Pattern.compile("(\\d+)\\.(\\d+)\\.(\\d+)([\\.-].*)?");
     private static Version version;
     private static String serverId = System.getProperty(KieServerConstants.KIE_SERVER_ID);
-    private static String name = System.getProperty(KieServerConstants.KIE_SERVER_ID);
+    private static String name;
     private static String contextRoot;
 
     static {
+    	if (serverId == null || serverId.length() == 0) {
+    		serverId = System.getenv("KIE_SERVER_ID");
+        	if (serverId == null || serverId.length() == 0) 
+        		throw new IllegalArgumentException("Missing -Dorg.kie.server.id or env:KIE_SERVER_ID");
```

Comment:
In case serverId is null then it is generated by https://github.com/kiegroup/droolsjbpm-integration/blob/master/kie-server-parent/kie-server-services/kie-server-services-common/src/main/java/org/kie/server/services/Bootstrap.java#L41-L47 .
Server id isn't mandatory parameter. Please don't throw an exception in this case.

Reply:
Committed changes as per requested

Reply:
I'm not sure I follow. If serverId is guaranteed to never be null, there is no harm in throwing an exception, right? If it might be null, is there now a problem down the line, with a null name?

Reply:
**_It's all about timing_**
There are 3 stages to resolve and valid KieServerId. 

**Stage 1** - At class loading time, the static initializer of KieServerEnvironment class will attempt to load it from '-D' java property first and followed by associated environment variable. Initially, a check has been implemented here but soon we have realized that it may break existing code which may NOT require an explicit configuration of KieServerId. (KieServerId may be generated at stage 2). As a result, we have removed and postponed such logic to later stage (At stage 3).

**Stage 2** - At KieServer WebApp bootstrap time, the ServletContextListener callback generates a random KieServerId as pointed out by Karel. However, this is not compatible with new code. To resolve such issue. The validation logic has been implemented at Stage 3 within KieServerStateCloudRepository.load method.

**Stage 3** - During KieServer initialization period, KieServerImpl loads and stores KieServerState several times. An exception will be thrown if a generated KieServerId has somehow being used. (See below) 

https://github.com/rhtevan/droolsjbpm-integration/blob/b2c6f68b6dfb850f26726796d29ea877f743efd4/kie-server-parent/kie-server-services/kie-server-services-common/src/main/java/org/kie/server/services/impl/storage/cloud/KieServerStateCloudRepository.java#L141-L145 

In short, KieServerStateCloudRepository and OpenShiftStartupStrategy require a legitimate KieServerId being passed via '-D' or ENV variable. By moving the check to the last stage will eliminate any impact to existing startup strategy and file based state repository.